### PR TITLE
Bug fixes to ensure unique particle ids in uniform_box.cc and uniform_radial.cc

### PR DESCRIPTION
**Pull request by egpuckett**
Rene fixed this bug in uniform_radial.cc at the hackathon last May (https://github.com/EGP-CIG-REU/aspect/commit/c790d018338dcae83752c39356a0a296f6b2e98f#diff-3019d3e12a3ca9578bd44924a184f935) but somehow it was lost when merging the tracer-improvements branch into the master last fall. Uniform_box.cc never had this bug fix implemented. 

I believe the other files that have been changed in my master w.r.t. the upstream master were changed when I ran astyle on my code tree.

### CODE REVIEW

File: source/particle/generator/uniform_box.cc [hunk 1]
```diff
@@ -79,8 +80,11 @@ namespace aspect
                       // worry about it and move on to next point.
                       try
                         {
-                          particles.insert(this->generate_particle(particle_position,particle_index));
+                          // Moved particle_index++ ABOVE the call to particles.insert in
+                          // order to ensure that all particles generated on all processers
+                          // have a unique particle id (i.e., particle_index)
```
Comment (bangerth) [inline source/particle/generator/uniform_box.cc]: Let's remove this comment. When reading code, what matters only how it looks *now*, not through which process we *arrived* here.

File: source/particle/generator/uniform_radial.cc [hunk 2]
```diff
@@ -80,13 +80,15 @@ namespace aspect
...
                     const Point<dim> particle_position = Utilities::cartesian_coordinates<dim>(spherical_coordinates) + P_center;
-
                     // Try to add the particle. If it is not in this domain, do not
                     // worry about it and move on to next point.
                     try
                       {
-                        particles.insert(this->generate_particle(particle_position,particle_index));
+                        // Moved particle_index++ ABOVE the call to particles.insert in
+                        // order to ensure that all particles generated on all processers
+                        // have a unique particle id (i.e., particle_index)
```
Comment (bangerth) [inline source/particle/generator/uniform_radial.cc]: same here

File: include/aspect/assembly.h [hunk 3]
```diff
@@ -377,7 +377,7 @@ namespace aspect
          * - The copy object into which assemblers add up their contributions.
          */
         boost::signals2::signal<void (const double,
-                                      internal::Assembly::Scratch::StokesPreconditioner<dim>  &,
+                                      internal::Assembly::Scratch::StokesPreconditioner<dim> &,
```
Comment (bangerth) [inline include/aspect/assembly.h]: These changes are a bit of a shame since we had so nicely aligned the ampersands of arguments.

Could you limit the patch to only those files that really have functional changes?

**Comment by bangerth** [comment]
I think that if you ran the indentation script, that most of the whitespace changes would disappear again.

@gassmoeller -- The changes seem reasonable to me but I'll leave the call to you since you have a better overview.

File: source/particle/generator/uniform_box.cc (hunk 1 above)
Comment (egpuckett) [inline source/particle/generator/uniform_box.cc]: Will do!

File: source/particle/generator/uniform_radial.cc (hunk 2 above)
Comment (egpuckett) [inline source/particle/generator/uniform_radial.cc]: Also ...

File: include/aspect/assembly.h (hunk 3 above)
Comment (egpuckett) [inline include/aspect/assembly.h]: Which version of astyle are you using.  I have 2.03, which indent refuses to run with.  It says 2.04 is required, but here are now:

astyle 2.05.1	2014-12-11	

astyle 2.05	2014-11-17

astyle 2.04	2013-11-01

Comment (egpuckett) [inline include/aspect/assembly.h]: I figured it out: astyle.rc insists on version 2.04

Comment (bangerth) [inline include/aspect/assembly.h]: You need precisely 2.04.

**Comment by egpuckett** [comment]
I downloaded and installed astyle version 2.04. Then I ran doc/indent. Now astyle tester is 'OK'.

File: source/particle/generator/uniform_radial.cc [hunk 4]
```diff
@@ -80,13 +80,12 @@ namespace aspect
                   {
                     spherical_coordinates[1] = P_min[1] + j * phi_spacing;
                     const Point<dim> particle_position = Utilities::cartesian_coordinates<dim>(spherical_coordinates) + P_center;
-
```
Comment (bangerth) [inline source/particle/generator/uniform_radial.cc]: unrelated change -- can you undo this one?

File: include/aspect/assembly.h [hunk 5]
```diff
@@ -398,8 +398,8 @@ namespace aspect
         boost::signals2::signal<void (const typename DoFHandler<dim>::active_cell_iterator &,
                                       const double,
                                       const bool,
-                                      internal::Assembly::Scratch::StokesSystem<dim>       &,
-                                      internal::Assembly::CopyData::StokesSystem<dim>      &)> local_assemble_stokes_system;
+                                      internal::Assembly::Scratch::StokesSystem<dim> &,
+                                      internal::Assembly::CopyData::StokesSystem<dim> &)> local_assemble_stokes_system;
```
Comment (bangerth) [inline include/aspect/assembly.h]: here, and in line 427, this is an unrelated change. can you restore the original indentation?

(I say these things not because I'm super pedantic -- which is also true -- but primarily because with many others developing code in this area right now, anyone who changes lines and gets them merged may cause merge conflicts for others who also made changes in this area but did not get them merged yet. It's therefore always useful to avoid unnecessary changes in unrelated areas to avoid stepping on other developers' feet.)

Comment (bangerth) [inline include/aspect/assembly.h]: and in line 380 above.

Comment (egpuckett) [inline include/aspect/assembly.h]: Hi Wolfgang, I'm rather pedantic myself and you've given an excellent reason for keeping the alignment uniform.  Harsha and I will take care of this Friday.  BTW, all of these changes were made by astyle and indent, although I realize at first I was using astyle 2.03.

**Comment by gassmoeller** [comment]
This indeed solves the problem of #770, however, it also changes the behavior of the generators, as you can see in the failing tests. The particle IDs now start from 1 for these generators, while they start from 0 for all other generators. I would prefer a slight modification: If you move the line `particle_index++;` behind the try-catch block, you should achieve the same behavior, while still solving the problem (increase the ID for every global position, but only create the particle if it is in the local domain). Would that work?
With the given modification and if you revert the remaining unrelated changes commented by Wolfgang this looks good. 
Let me know when you are done with the changes and I will give it another look.

**Comment by egpuckett** [comment]
Hi Rene, Harsha and I have (hopefully) resolved the formatting issues by making the changes by hand and not running astyle and indent.  We also moved the line `particle_index++;` behind the try-catch block as you suggested.  We're testing it here now.

File: include/aspect/assembly.h [hunk 6]
```diff
@@ -398,8 +398,8 @@ namespace aspect
         boost::signals2::signal<void (const typename DoFHandler<dim>::active_cell_iterator &,
                                       const double,
                                       const bool,
-                                      internal::Assembly::Scratch::StokesSystem<dim>       &,
-                                      internal::Assembly::CopyData::StokesSystem<dim>      &)> local_assemble_stokes_system;
+                                      internal::Assembly::Scratch::StokesSystem<dim>      &,
+                                      internal::Assembly::CopyData::StokesSystem<dim>     &)> local_assemble_stokes_system;
```
Comment (gassmoeller) [inline include/aspect/assembly.h]: very close, but you need to add one more space in both lines

File: include/aspect/assembly.h [hunk 7]
```diff
@@ -424,8 +424,8 @@ namespace aspect
                                       const unsigned int,
                                       const double,
                                       const bool,
-                                      internal::Assembly::Scratch::StokesSystem<dim>       &,
-                                      internal::Assembly::CopyData::StokesSystem<dim>      &)> local_assemble_stokes_system_on_boundary_face;
+                                      internal::Assembly::Scratch::StokesSystem<dim>      &,
+                                      internal::Assembly::CopyData::StokesSystem<dim>     &)> local_assemble_stokes_system_on_boundary_face;
```
Comment (gassmoeller) [inline include/aspect/assembly.h]: Same here, one more space

File: source/particle/generator/uniform_box.cc [hunk 8]
```diff
@@ -64,11 +64,12 @@ namespace aspect
                     // worry about it and move on to next point.
                     try
                       {
+
```
Comment (gassmoeller) [inline source/particle/generator/uniform_box.cc]: please remove this empty line

**Comment by gassmoeller** [comment]
There are just a few very minor things left (see comments above). Lets see what the tester says.

**Comment by egpuckett** [comment]
I think I made the formatting changes as you requested. What troubles me is how do I get the format correct in the future.  Shouldn't running astyle 2.04 be sufficient?

**Comment by gassmoeller** [comment]
Ok looks good, thanks! 
One last thing to do: Could you add an entry to doc/modules/changes.h in which you mention what you fixed, and who fixed it (just copy one of the existing entries and add it to the top of the list). E.g. 'Fixed: The uniform radial and uniform box particle generators now produce globally unique particle IDs.'

Running astyle 2.04 is in almost all cases sufficient. The spacing between variable type and name is one of the very few things that is not corrected by astyle. But this is only an issue if you change the spacing in existing code, in new code you are free to choose this spacing in a reasonable way yourself. Almost all other things are corrected automatically by astyle.

**Comment by egpuckett** [comment]
Hi Rene, I have added the entry to doc/modules/changes.h as you requested. I think I kept the format of that file as it should be, but someone should double check.  Thanks for your help! - G

**Comment by gassmoeller** [comment]
Ok, entry looks good. Thanks for the fix!
